bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+]
+"""
+
+from __future__ import annotations
+from dataclasses import dataclass
+from difflib import SequenceMatcher
+
+import json
+import os
+import requests
+import sys
+import tomllib
+
+from http.client import HTTPSConnection, HTTPResponse
+from typing import (
+    Any,
+    Dict,
+    List,
+    Generic,
+    Optional,
+    Tuple,
+    NamedTuple,
+    Set,
+    Iterable,
+    Protocol,
+    TypeVar,
+)
+class Member(NamedTuple):
+    id: str
+    name: str
+    email: str
+    type: str
+    access_all: Optional[bool]
+    # groups: Tuple[str, ...]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Member:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+        return Member(
+            id=data["member_id"],
+            name=data["member_name"],
+            email = data["email"],
+            type = data["type"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[member]]",
+            f"member_id = {self.id}",
+            f"member_name = {self.name}",
+            f"email = {self.email}",
+            f"type = {str(self.type)}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+class GroupMember(NamedTuple):
+    member_id: int
+    member_name: str
+    group_name: str
+
+    def get_id(self) -> str:
+        # Our generic differ has the ability to turn add/removes into changes
+        # for pairs with the same id, but this does not apply to memberships,
+        # which do not themselves have an id, so the identity to group on is
+        # the value itself.
+        return f"{self.member_id}@{self.group_name}"
+
+    def format_toml(self) -> str:
+        # Needed to satisfy Diffable, but not used in this case.
+        raise Exception(
+            "Group memberships are not expressed in toml, "
+            "please print the diffs in some other way."
+        )
+
+class Group(NamedTuple):
+    id: str
+    name: str
+    access_all: Optional[bool]
+
+    def get_id(self) -> str:
+        return self.id
+
+    @staticmethod
+    def from_toml_dict(data: Dict[str, Any]) -> Group:
+        access_all: Optional[bool] = False
+        if "access_all" in data:
+            access_all = data["access_all"]
+
+        return Group(
+            id=data["group_id"],
+            name=data["group_name"],
+            access_all = access_all,
+        )
+
+    def format_toml(self) -> str:
+        lines = [
+            "[[group]]",
+            f"group_id = {self.id}",
+            f"group_name = {self.name}",
+            f"access_all = {str(self.access_all).lower()}",
+        ]
+        return "\n".join(lines)
+
+
+class MemberCollectionAccess(NamedTuple):
+    name: str
```

Comment:
If it has only the member name in it, then we might as well use the member name instead of this wrapper object in places where it’s needed.

For GitHub, we also have the id in there because on a repository you can have a collaborator who is not a member of your organization (and username is not stable on GitHub, you can change it, so you can’t use it to identify accounts). So then the user is not recorded in the organization members, so we need to specify the details directly on the repository access settings. But in Bitwarden, I don’t think it is possible to give access to an account who is not part of the organization, so the username should reference somebody that’s already part of the organization, and we don’t need the id.

Reply:
Yes it was here because at the origin I was storing `id` (not needed as you explained) and the `group` the user belong to:

```
[[collection]]
collection_id = "8e69ce49-85ae-4e09-a52c-afaf00a90a3f"
external_id = ""
member_access = [
  { member_id = "2564c11f-fc1b-4ec7-aa0b-afaf00a9e4a4", member_name = "yan", group = "c6a13b93-edc1-4c3b-9fc5-afaf00a8d33f" },
]
```

We would keep it because we will store the `access` of the member to the `collection` which can be different on every collection ([see your comment](https://github.com/ChorusOne/github-access-manager/pull/10#discussion_r1120538272))

You mention that we don't need the id for the MemberAccessCollection and you're explanation makes sense to me, therefore I will ignore your comment `# Probably need to have the user id in there too.`
```
# This lists only _direct_ access, i.e. not members that have access through being member of a group.
member_access = [
  { member_name = "yan", access="write" }, # Probably need to have the user id in there too
]
```

Reply:
> therefore I will ignore your comment `# Probably need to have the user id in there too.`

:+1: Yes makes sense, if all members are part of the organization, their ids are recorded elsewhere.